templates/mip/main.go

```diff
@@ -0,0 +1,137 @@
+// package main holds the implementation of the mip-knapsack template.
+package main
+
+import (
+	"context"
+	"log"
+
+	"github.com/nextmv-io/sdk/mip"
+	"github.com/nextmv-io/sdk/run"
+	"github.com/nextmv-io/sdk/run/schema"
+)
+
+// This template demonstrates how to solve a Mixed Integer Programming problem.
+// To solve a mixed integer problem is to optimize a linear objective function
+// of many variables, subject to linear constraints. We demonstrate this by
+// solving the well known knapsack problem.
+func main() {
+	err := run.CLI(solver).Run(context.Background())
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+// The options for the solver.
+type options struct {
+	Limits mip.Limits `json:"limits,omitempty"`
+}
+
+// Input of the problem.
+type input struct {
+	Items          []item  `json:"items"`
+	WeightCapacity float64 `json:"weight_capacity"`
+}
+
+// An item has a Value and Weight. ItemID is used to identify the item.
+type item struct {
+	ItemID string  `json:"item_id,omitempty"`
+	Value  float64 `json:"value"`
+	Weight float64 `json:"weight"`
+}
+
+// ID is implemented to fulfill the model.Identifier interface.
+func (i item) ID() string {
+	return i.ItemID
+}
+
+// solution represents the decisions made by the solver.
+type solution struct {
+	Items []item `json:"items,omitempty"`
+}
+
+func solver(_ context.Context, input input, options options) (schema.Output, error) {
+	// We start by creating a MIP model.
+	m := mip.NewModel()
+
+	// Create a map of ID to decision variables for each item in the knapsack.
+	itemVariables := make(map[string]mip.Bool, len(input.Items))
+	for _, item := range input.Items {
+		// Create a new binary decision variable for each item in the knapsack.
+		itemVariables[item.ItemID] = m.NewBool()
+	}
+
+	// We want to maximize the value of the knapsack.
+	m.Objective().SetMaximize()
+
+	// This constraint ensures the weight capacity of the knapsack will not be
+	// exceeded.
+	capacityConstraint := m.NewConstraint(
+		mip.LessThanOrEqual,
+		input.WeightCapacity,
+	)
+
+	// For each item, set the term in the objective function and in the
+	// constraint.
+	for _, item := range input.Items {
+		// Sets the value of the item in the objective function.
+		m.Objective().NewTerm(item.Value, itemVariables[item.ItemID])
+
+		// Sets the weight of the item in the constraint.
+		capacityConstraint.NewTerm(item.Weight, itemVariables[item.ItemID])
+	}
+
+	// Create a solver using a provider. Please see the documentation on
+	// [mip.SolverProvider] for more information on the available providers.
+	solver, err := mip.NewSolver(mip.Highs, m)
```

Comment:
If we keep the comment from the yaml referring to xpress, we should potentially point out that the user also needs to modify this line according to the solver they are choosing. :shrug:

Reply:
I'll leave this as is for it is fairly generic to support more solvers. I updated the comment in the yaml.